### DISCUSSION TITLE
Initial experience using dotNetRdf with Godot

### DISCUSSION THREAD

**Discussion by calummackervoy** [dotnetrdf/dotnetrdf#483] [General]
Hi! I've been experimenting for a little while with building a tool which allows games to share characters, world data and features, using RDF technologies, Shapes and Solid

I wanted to build something with some 2D graphics, so I started working on [a project using Godot](https://github.com/calummackervoy/SimPolis). I've worked with RDF in the past (using Jena, PyLD, a little with RDFLib and with RDF.JS), I've done a little Godot, but C# and this library are both new to me

The first thing I noticed is that the .NET versionings are very confusing. There is Mono, Core, SDK, runtime, framework and this is quite overwhelming for a beginner. Godot aren't very clear which versions they exactly support, for example C# in Godot 3.2.3 (Mono) has a bundled SDK but requires installing .NET Core SDK

I followed [Godots guide](https://docs.godotengine.org/en/stable/tutorials/scripting/c_sharp/c_sharp_basics.html) on getting setup, from here I found that it was easy to install dotNetRDF using the documentation and getting the Hello world application to run. Reading RDF data from the web has been a different story, and I haven't been able to get any combination to work involving a HTTP request. At the moment I'm wondering if there's some kind of versioning issue in my setup, because the problems I'm having seem off (like not being able to find a parser for `application/ld+json`, and not being able to read turtle behind a HTTPS request)

If anyone has a pointer on this which might help me then it'll be much appreciated, but I suppose my main reason for posting is because my use case might be interesting. For now I might wait for Godot 4 which promises better C# support, and shift the demo features I need onto the server-side in the time being (most likely using Python and Flask)

**Comment by kal** [comment]
Hi Calum,

Thanks for the input. I've only been vaguely aware of Godot so in that sense I think you may be breaking some new ground using dotNetRDF and Godot together, but I'm happy to help if you hit specific issues. I'm glad that you found the docs helpful in getting started, but also I'd be keen to learn more about where you got stuck to see if there are ways in which the docs can be improved. This is especially true now as the upcoming 3.0 release changes quite a lot of APIs and so the docs will need a complete refresh in any case so it is a good time to incorporate new ideas and suggestions.

With regards to the `application/ld+json`, any 2.x version of dotNetRDF has support for JSON-LD (2.6 or later bumps support to JSON-LD 1.1). Are you using the `UriLoader` static class for this or processing an web response yourself? `UriLoader` should be able to detect the format as long as the response header specifies `application/ld+json` in the `Content-Type` header (if the response specifies `appilcation/json` or `text/json` then that will be treated as being RDF/JSON which is a different format and will fail to parse).

With regards to HTTPS requests, this *may* be because of the way that .NET Core handles HTTPS requests. By default I believe support for TLS 1.1 and 1.2 are *not* enabled and so you need a bit of startup code to do that:

```c#
using System.Net;
...
System.Net.ServicePointManager.SecurityProtocol |= SecurityProtocolType.Tls11 | SecurityProtocolType.Tls12;
```

If you are still having trouble, feel free to either open a request here or on StackOverflow and I'll do my best to help. Any code snippets you can share would be helpful too so that I can see which APIs you are using as dotNetRdf does provide multiple different ways to interact with Linked Data and its not always obvious from a question which particular approach a developer is using.

**Reply by calummackervoy** [reply]
Hi! Thanks for the response 😄 I'm happy to try and get through this and hopefully it will be useful for the library - I'm aware my use case is quite niche. Since Godot is heading in a C# direction it might be less niche in the future 🤞 

Since writing this I've found a winning formula (which I got from [this Stack Overflow question](https://stackoverflow.com/questions/62233728/how-to-parse-json-ld-feed-using-dotnetrdf-in-c-sharp))

```
var g = new TripleStore();
g.LoadFromUri(new Uri("http://localhost:5000/get-static-demo-map/"));
```

I'll open a few issues for the other solutions I tried. Happy to help with debugging - I don't know if it's just a case of my misuse of the library but maybe it will help with documentation if so
